packages/itwinui-react/src/core/ComboBox/ComboBox.tsx

```diff
@@ -412,33 +419,42 @@ export const ComboBox = React.forwardRef(
           return;
         }
 
-        if (isMultipleEnabled(selected, multiple)) {
+        if (isMultipleEnabled(value, multiple)) {
           const actionType = isMenuItemSelected(__originalIndex)
             ? 'removed'
             : 'added';
-          const newArray = selectedChangeHandler(__originalIndex, actionType);
-          dispatch({ type: 'multiselect', value: newArray });
+          const newArray = selectedChangeHandler(
+            __originalIndex,
+            actionType,
+            value,
+          );
+          setValue(newArray);
           onChangeHandler(__originalIndex, actionType, newArray);
 
           // update live region
           setLiveRegionSelection(
             newArray
-              .map((item) => optionsRef.current[item]?.label)
+              .map(
+                (item) =>
+                  optionsRef.current.find((option) => option.value === item)
```

Comment:
We should undo all the changes regarding `value` vs `index`. It's not related to the controlled state bug, and it can degrade performance. I believe using `index` was a deliberate design decision to avoid having to repeatedly call `find`/`findIndex`.

Reply:
Ohh, I see. Reverted back to using index over value (https://github.com/iTwin/iTwinUI/pull/2048/files/ff7b08b79d001463af7c3f636bac854ca8d234fa..e9b231507291fa11e7b5f305a2bca08e831c1c5e). Since this was a big change, also updated the PR description significantly.